qiskit/namespace.py

```diff
@@ -0,0 +1,69 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+# pylint: disable=unused-argument
+
+"""Module for utilities to manually construct qiskit namespace"""
+
+import sys
+from importlib.abc import MetaPathFinder, Loader
+import importlib
+
+
+def _new_namespace(fullname, old_namespace, new_package):
+    names = fullname.split(".")
+    new_namespace_names = new_package.split(".")
+    old_namespace_names = old_namespace.split(".")
+    fullname = ".".join(new_namespace_names + names[len(old_namespace_names) :])
+    return fullname
+
+
+class QiskitLoader(Loader):
+    """Load qiskit element as a namespace package."""
+
+    def __init__(self, new_package, old_namespace):
+        super().__init__()
+        self.new_package = new_package
+        self.old_namespace = old_namespace
+
+    def module_repr(self, module):
+        return repr(module)
+
+    def load_module(self, fullname):
+        old_name = fullname
+        fullname = _new_namespace(fullname, self.old_namespace, self.new_package)
+        module = importlib.import_module(fullname)
+        sys.modules[old_name] = module
+        return module
+
+
+class QiskitElementImport(MetaPathFinder):
+    """Meta importer to enable unified qiskit namespace."""
+
+    def __init__(self, old_namespace, new_package):
+        super().__init__()
+        self.old_namespace = old_namespace
+        self.new_package = new_package
+
+    def find_spec(self, fullname, path=None, target=None):
+        """Return the ModuleSpec for Qiskit element."""
+        if fullname.startswith(self.old_namespace):
```

Comment:
I suppose technically this might break some compatibility if some random package is extending the namespace with `qiskit.providers.aer_with_my_modifications` (still `startswith("qiskit.providers.aer")`, but we probably don't need to care about that.

Reply:
I'm not sure namespace packaging would work for someone trying to extend `qiskit.providers.aer` since aer itself doesn't add the same pkgutil hooks. In some cases someone probably could extend the namespace with their custom subpackage but not in all permutations of packaging as namespace packaging isn't reliable (this is why I couldn't remove the pkgutil calls otherwise I'd break compat with old/current aer releases).

Reply:
I mean if somebody extends `qiskit.providers` with a name that starts with `aer`, like `aer_but_actually_not`.  I think it's probably not worth worrying about, though.

Reply:
ah ok now I follow what you're saying. Yeah I should have done a proper regex match something like `^qiskit\.provider\.aer\.*` to match on exactly the aer namespace. But yeah I'm not too worried a potential name squatter like that. The bigger concern for me is how we deprecate and remove the pkgutil namespace hooks longer term. I know there are external users of it and trying to migrate them off will be a pain.